fs/bcachefs/checksum.c

```diff
@@ -16,53 +20,115 @@
 #include <crypto/skcipher.h>
 #include <keys/user-type.h>
 
-static u64 bch2_checksum_init(unsigned type)
+/*
+ * bch2_checksum state is an abstraction of the checksum state calculated over different pages.
+ * it features page merging without having the checksum algorithm lose its state.
+ * for native checksum aglorithms (like crc), a default seed value will do, and those can easily picked up where they left off.
+ * for hash-like algorithms, a state needs to be stored
+ */
+
+struct bch2_checksum_state {
+	union {
+		u64 seed;
+#ifdef CONFIG_BCACHEFS_ADDITIONAL_HASHES
+		struct xxh64_state h64state;
+#endif
+	};
+	unsigned type;
+};
+
+static void bch2_checksum_init(struct bch2_checksum_state* state)
 {
-	switch (type) {
+	switch (state->type) {
 	case BCH_CSUM_NONE:
-		return 0;
-	case BCH_CSUM_CRC32C_NONZERO:
-		return U32_MAX;
-	case BCH_CSUM_CRC64_NONZERO:
-		return U64_MAX;
 	case BCH_CSUM_CRC32C:
-		return 0;
 	case BCH_CSUM_CRC64:
-		return 0;
+		state->seed = 0;
+		break;
+	case BCH_CSUM_CRC32C_NONZERO:
+		state->seed = U32_MAX;
+		break;
+	case BCH_CSUM_CRC64_NONZERO:
+		state->seed = U64_MAX;
+		break;
+#ifdef CONFIG_BCACHEFS_ADDITIONAL_HASHES
+	case BCH_CSUM_XXHASH:
+		xxh64_reset(&state->h64state, 0);
+		break;
+	case BCH_CSUM_XOR_BLOCK:
+		state->seed = 0;
+		break;
+#endif
 	default:
 		BUG();
 	}
 }
 
-static u64 bch2_checksum_final(unsigned type, u64 crc)
+static u64 bch2_checksum_final(struct bch2_checksum_state* state)
 {
-	switch (type) {
+	switch (state->type) {
 	case BCH_CSUM_NONE:
-		return 0;
-	case BCH_CSUM_CRC32C_NONZERO:
-		return crc ^ U32_MAX;
-	case BCH_CSUM_CRC64_NONZERO:
-		return crc ^ U64_MAX;
 	case BCH_CSUM_CRC32C:
-		return crc;
 	case BCH_CSUM_CRC64:
-		return crc;
+		return state->seed;
+#ifdef CONFIG_BCACHEFS_ADDITIONAL_HASHES
+	case BCH_CSUM_XXHASH:
+		return xxh64_digest(&state->h64state);
+	case BCH_CSUM_XOR_BLOCK:
+		return state->seed;
+#endif
+	case BCH_CSUM_CRC32C_NONZERO:
+		return state->seed ^ U32_MAX;
+	case BCH_CSUM_CRC64_NONZERO:
+		return state->seed ^ U64_MAX;
 	default:
 		BUG();
 	}
 }
 
-static u64 bch2_checksum_update(unsigned type, u64 crc, const void *data, size_t len)
+#ifdef CONFIG_BCACHEFS_ADDITIONAL_HASHES
+
+static void bch2_xor_block_helper(struct bch2_checksum_state* state, const u64* data, const size_t u64_len) {
```

Comment:
Is `u64_len` guaranteed to be less than the max xor_blocks source count size? `xor_blocks` takes a max of 4 for the source count right?

Reply:
seems you are right.  I thought the xor_blocks function would contain a for loop to iterate over count size, but this seems wrong.  I'll adapt it

Reply:
fixed